### ISSUE TITLE
Configurations with circular references cause Karma to fail to run

### ISSUE THREAD

**Issue by filipesilva** [karma-runner/karma#3154]
### Expected behaviour
Karma runs even if there are circular references in the configuration. 

These can happen with custom keys added directly to the configuration, or inside plugin modules. The internal data structure should not cause Karma to fail to run.

### Actual behaviour
Karma fails to run with `TypeError: Converting circular structure to JSON`.

### Environment Details

- Karma version (output of `karma --version`): 3.0.0
- Relevant part of your `karma.config.js` file
```
const selfRefObject = {};
selfRefObject.ref = selfRefObject;

module.exports = function (config) {
  config.set({
    basePath: '',
    frameworks: ['jasmine'],
    plugins: [
      require('karma-jasmine'),
      require('karma-chrome-launcher'),
    ],
    browsers: ['ChromeHeadless'],
    files: ['./simple.spec.js'],
    selfRefObject
  });
};
```

### Steps to reproduce the behaviour

1. `git clone https://github.com/filipesilva/karma-circular-structure`
2. `npm install`
3. `npm test`

This error happens because `JSON.stringify` is called on the final configuration in these two lines:
https://github.com/karma-runner/karma/blob/20eab32345a2e47334aada35ab6cf7c0a0000b7b/lib/server.js#L65

https://github.com/karma-runner/karma/blob/20eab32345a2e47334aada35ab6cf7c0a0000b7b/lib/server.js#L363

Since the configuration contains arbitrary data, objects, modules, and functions, `JSON.stringify` might not be the best way way of serializing it.

Related to https://github.com/karma-runner/karma/issues/3075

**Comment by djusv** [comment]
I had the similar error when I was using karma-webpack. When I had ```ForkTsCheckerWebpackPlugin``` defined in the plugins of my webpack configuration I had the same problem. My workaround was to remove this plugin from the Webpack Test Configuration.

**Comment by filipesilva** [comment]
@djusv as a workaround you can also add a `toJSON: () => { },` method to the problematic configuration object.

In the repro above, it would be:
```
const selfRefObject = {};
selfRefObject.ref = selfRefObject;
selfRefObject.toJSON = () => { };
```

When that method is present, `JSON.stringify()` will use it instead of parsing the object.

**Comment by avantika-karwasara** [comment]
karma=> 3.1.1
getting this error 
`Converting circular structure to JSON`
Is this error  fixed in any version?

**Comment by devoto13** [comment]
Closing as this issue has been fixed in [3.1.2](https://github.com/karma-runner/karma/blob/master/CHANGELOG.md#312-2018-12-01) and then improved in [4.2.0](https://github.com/karma-runner/karma/blob/master/CHANGELOG.md#420-2019-07-12)
